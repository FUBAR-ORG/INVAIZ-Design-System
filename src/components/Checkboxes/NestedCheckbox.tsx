import type { PropsWithChildren, ComponentProps } from 'react';
// types

import styled from '@emotion/styled';
// React modules

import OutlineCheckbox, { ParentCheckbox } from '@components/Checkboxes/OutlineCheckbox';
// components

/**
 * 중첩으로 계층 구조를 표현할 수 있는 체크박스 아이템을 포함하고 있는 컴포넌트입니다.
 *
 * * 아이템 요소를 `NestedCheckbox.Item`과 같이 사용할 수 있는 `Compound Pattern`을 사용합니다.
 */
function NestedCheckbox({
  children,
  ...props
}: PropsWithChildren<ComponentProps<typeof ParentCheckbox>>) {
  return (
    <div>
      <ParentCheckbox {...props} />
      <StyleNestedCheckList>{children}</StyleNestedCheckList>
    </div>
  );
}

const NestedCheckboxItem = ({ ...props }: ComponentProps<typeof OutlineCheckbox>) => (
  <StyleNestedCheckItem>
    <OutlineCheckbox {...props} />
  </StyleNestedCheckItem>
);

/**
 * 중첩 계층 구조를 표현할 때 사용하는 체크박스 아이템 컴포넌트입니다.
 */
NestedCheckbox.Item = NestedCheckboxItem;

const StyleNestedCheckList = styled.ul`
  list-style: none;

  margin: -2px 0 0 10px;
  padding: 0 0 0 20px;

  & > li + div {
    margin-top: -10px;
  }
`;

const StyleNestedCheckItem = styled.li`
  margin-top: -10px;
`;

export default NestedCheckbox;
