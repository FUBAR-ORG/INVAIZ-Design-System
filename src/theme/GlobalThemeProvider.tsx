import { PropsWithChildren } from 'react';
import { ThemeProvider } from '@emotion/react';
import font, { fontSize, fontWeight, lineHeight } from '@theme/font';
import color from '@theme/color';

export default function GlobalThemeProvider({ children }: PropsWithChildren<unknown>) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

const theme = { font, fontSize, lineHeight, fontWeight, color };
