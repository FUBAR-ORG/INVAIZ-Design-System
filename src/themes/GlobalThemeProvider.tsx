import type { PropsWithChildren } from 'react';
import type { EmotionTheme } from '@themes/emotion-theme';
import type { ColorTheme } from '@themes/colors/color';
// types

import { ThemeProvider } from '@emotion/react';
// React modules

import font, { fontSize, fontWeight, lineHeight, textAlign, fontColor } from '@themes/font';
import { colorMap, normal, light } from '@themes/colors/color';
import style from '@themes/style';
import GlobalStyle from '@themes/GlobalStyle';

interface GlobalThemeProviderProps {
  colorTheme?: ColorTheme;
}

export default function GlobalThemeProvider({
  children,
  colorTheme = 'normal',
}: PropsWithChildren<GlobalThemeProviderProps>) {
  const theme: EmotionTheme = {
    normal,
    light,

    color: colorMap[colorTheme],

    font,
    fontSize,
    lineHeight,
    fontWeight,
    textAlign,
    fontColor,

    style,
  };
  return (
    <ThemeProvider theme={theme}>
      {children}
      <GlobalStyle />
    </ThemeProvider>
  );
}
