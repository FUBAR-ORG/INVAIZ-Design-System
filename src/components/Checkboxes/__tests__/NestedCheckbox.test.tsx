import { create } from 'react-test-renderer';
// test utils

import NestedCheckbox from '@components/Checkboxes/NestedCheckbox';
// components

import GlobalThemeProvider from '@themes/GlobalThemeProvider';
// providers

const PARENT_TEXT = 'Parent Checkbox' as const;
const CHILDREN_TEXT = 'Children' as const;
// constant

test('NestedCheckbox Snapshot', () => {
  const indeterminate = create(
    <GlobalThemeProvider>
      <NestedCheckbox text={PARENT_TEXT} isIndeterminate>
        <NestedCheckbox.Item text={CHILDREN_TEXT} />
      </NestedCheckbox>
    </GlobalThemeProvider>
  );
  expect(indeterminate).toMatchSnapshot();

  const determinate = create(
    <GlobalThemeProvider>
      <NestedCheckbox text={PARENT_TEXT}>
        <NestedCheckbox.Item text={CHILDREN_TEXT} />
      </NestedCheckbox>
    </GlobalThemeProvider>
  );
  expect(determinate).toMatchSnapshot();
});
