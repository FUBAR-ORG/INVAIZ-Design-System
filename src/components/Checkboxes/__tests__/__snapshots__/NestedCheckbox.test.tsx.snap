// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedCheckbox Snapshot 1`] = `
<div>
  <label
    className="css-9ncwx"
  >
    <span
      className="css-1u8zmlt"
    >
      <input
        className="css-18gtr08"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="css-3xhjcj"
      >
        <svg
          className="visible css-bw5rgr"
          data-file-name="Subtract.svg"
          data-svg-name="Subtract"
          data-test-id="Subtract"
          height={20}
          width={20}
        />
      </div>
    </span>
    <p
      className="css-4ccycq"
    >
      Parent Checkbox
    </p>
  </label>
  <ul
    className="css-1tsmip4"
  >
    <li
      className="css-llbbcb"
    >
      <label
        className="css-9ncwx"
      >
        <span
          className="css-1u8zmlt"
        >
          <input
            className="css-18gtr08"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="css-3xhjcj"
          >
            <svg
              className="css-bw5rgr"
              data-file-name="Check.svg"
              data-svg-name="Check"
              data-test-id="Check"
              height={20}
              width={20}
            />
          </div>
        </span>
        <p
          className="css-4ccycq"
        >
          Children
        </p>
      </label>
    </li>
  </ul>
</div>
`;

exports[`NestedCheckbox Snapshot 2`] = `
<div>
  <label
    className="css-9ncwx"
  >
    <span
      className="css-1u8zmlt"
    >
      <input
        className="css-18gtr08"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        className="css-3xhjcj"
      >
        <svg
          className="css-bw5rgr"
          data-file-name="Check.svg"
          data-svg-name="Check"
          data-test-id="Check"
          height={20}
          width={20}
        />
      </div>
    </span>
    <p
      className="css-4ccycq"
    >
      Parent Checkbox
    </p>
  </label>
  <ul
    className="css-1tsmip4"
  >
    <li
      className="css-llbbcb"
    >
      <label
        className="css-9ncwx"
      >
        <span
          className="css-1u8zmlt"
        >
          <input
            className="css-18gtr08"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="css-3xhjcj"
          >
            <svg
              className="css-bw5rgr"
              data-file-name="Check.svg"
              data-svg-name="Check"
              data-test-id="Check"
              height={20}
              width={20}
            />
          </div>
        </span>
        <p
          className="css-4ccycq"
        >
          Children
        </p>
      </label>
    </li>
  </ul>
</div>
`;
